lib/AST/IRToASTVisitor.cpp

```diff
@@ -61,9 +63,86 @@ class ExprGen : public llvm::InstVisitor<ExprGen, clang::Expr *> {
   clang::Expr *visitUnaryOperator(llvm::UnaryOperator &inst);
 };
 
+clang::Expr *IRToASTVisitor::ConvertExpr(z3::expr expr) {
+  if (expr.decl().decl_kind() == Z3_OP_EQ) {
+    CHECK_EQ(expr.num_args(), 2) << "Equalities must have 2 arguments";
+    auto a{expr.arg(0)};
+    auto b{expr.arg(1)};
+
+    llvm::SwitchInst *inst{dec_ctx.z3_sw_vars_inv[a.id()]};
+    unsigned case_idx{};
+
+    if (!inst) {
+      inst = dec_ctx.z3_sw_vars_inv[b.id()];
+      case_idx = a.get_numeral_uint();
+    } else {
+      case_idx = b.get_numeral_uint();
+    }
+
+    for (auto sw_case : inst->cases()) {
+      if (sw_case.getCaseIndex() == case_idx) {
+        return ast.CreateEQ(CreateOperandExpr(inst->getOperandUse(0)),
+                            CreateConstantExpr(sw_case.getCaseValue()));
+      }
+    }
+
+    LOG(FATAL) << "Couldn't find switch case";
+  }
+
+  auto hash{expr.id()};
+  if (dec_ctx.z3_br_edges_inv.find(hash) != dec_ctx.z3_br_edges_inv.end()) {
+    auto edge{dec_ctx.z3_br_edges_inv[hash]};
+    CHECK(edge.second) << "Inverse map should only be populated for branches "
+                          "taken when condition is true";
+    return CreateOperandExpr(*(edge.first->op_end() - 3));
+  }
+
+  if (dec_ctx.z3_sw_vars_inv.find(hash) != dec_ctx.z3_sw_vars_inv.end()) {
+    auto inst{dec_ctx.z3_sw_vars_inv[hash]};
+    return CreateOperandExpr(inst->getOperandUse(0));
+  }
+
+  std::vector<clang::Expr *> args;
```

Comment:
Is this piece of code really necessary? It seems like the places where it's used we can easily just use `ConvertExpr(expr.arg(i))`, or am I missing something?

Reply:
Nope, taken out!